rust/datafusion/src/physical_plan/merge.rs

```diff
@@ -103,37 +105,56 @@ impl ExecutionPlan for MergeExec {
                 self.input.execute(0).await
             }
             _ => {
-                let tasks = (0..input_partitions).map(|part_i| {
+                let (sender, receiver) = mpsc::unbounded::<ArrowResult<RecordBatch>>();
```

Comment:
I wonder if picking a fixed channel size might be a good idea. Specifically, given there is s one  That way the producers can't get too far ahead. I think a small fixed number (`number_of_cores`? 10?) might be reasonable to begin with

Given that the actual work of `MergeStream` is pretty light (simply passing the record batches on) this may not be a problem in practice, but it probably depends on what order the tasks are run.

Reply:
Good point.

I actually started with a `bounded(1)`, but this (and others) won't work because we need to `join_all` threads. because there is no consumer to retrieve the items from the `receiver`, we are locked as the threads cannot be joined, and we wait indefinitely for the `join_all`.

An alternative is to not `join_all`, but then we risk losing results if the main thread finishes first.

I.e. if we bound the channel, we cannot `join_all` and thus we may lose threads. If we join all, the channel needs to be unbounded so that we can build the stream.

IMO neither is good, as in both cases we are essentially waiting for all threads to finish before returning the stream, which I understand is not what we want.

Reply:
Now that I think about it, I don't think calling `join_all` before returning `MergeStream` is needed here (as that  requires all tasks to complete executing before anything can consume the output of this operator).

If we want to pass errors back up stream (aka call join_all eventually) I wonder if we can put `handles` into `MergeStream` and call `join_all` after `input` is done. 

I am thinking all of this theoretically -- not having actually ever expressed such a thing in Rust / futures, BTW. So thanks for bearing with me!

Reply:
I thought that too!

until this morning, where I opened this PR and many tests failed. :/

The issue is not that we want the errors, is that we want to ensure the the threads finish. If we remove the `join_all`, we end up in the situation described here: https://stackoverflow.com/q/38957741/931303 (I think, thanks for bearing with me also ^_^)

Reply:
What experiment failed? removing the calls to `join_all`?

> is that we want to ensure the the threads finish

I wonder why we care about the thread "finishing" on some particular schedule. It seems what matter

The situation described in https://stackoverflow.com/q/38957741/931303 I think is about expecting to see a side-effect of thread execution that happens *after* all the results are retrieved from them. 

So in this case, the threads will all hang around, sending results via the channel as long as they have something to send and the channel hasn't been closed.

Once the channel is closed, there is some period of time during which the threads might be running (effectively after completing

Reply:
I'll try and give it look locally this afternoon. I am trying to finish up something for work now that I need to focus on first

Reply:
Thanks @alamb. So, [this](https://github.com/jorgecarleitao/arrow/pull/18) fails with no partitions collected at the end.

Reply:
I it that this is by design: there is no guarantee that a task spawn by Tokio finishes unless we `.join()` it. A `tokio::spawn` is a future like any other future: we need to await for it.

At least the problem is well defined: we have a vector (over `parts`) of futures (`spawn`) whose result is a stream (of `record batches`) and we would like to convert this into a stream of record batches. 

The solution IMO is to have a new adapter for this:

```rust
type Streams = Vec<Item=Stream<Item=Result<RecordBatch>>>;

struct Adapter {}

impl Adapter {
    pub fn new(it: Streams);
}

impl Stream for Adapter {
    item = Result<RecordBatch>
    
    fn poll_next(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Self::Item>> {
         todo!()
    }
}
```

Reply:
I am fairly convinced that the execution is not happening `async`hronously -- in particular, I think there is something about this code (or perhaps how some of the other collection functions) that means all data must be available *before* the stream future itself is returned. 

When I was playing with it, if `poll_next` ever returns `Pending` then `poll_next` is never called again. I think the adapter you outline above is a good idea; I think it also needs to arrange for `cx.waker()` to be woken if anything is added to the underlying stream.

I have a bunch of ideas in my head, but I have not had time to chase them down yet, sadly :(

Reply:
I have spent entirely too long on this today -- lol sorry it is taking so long.

I am now looking into the angle that this is related to how HashAggregate is implemented (specifically, it creates a stream and then calls `poll_next_unpin` on it: https://github.com/apache/arrow/blob/master/rust/datafusion/src/physical_plan/hash_aggregate.rs#L399). If the input is not all available immediately, I believe this stream is simply dropped and nothing ever drives the futures forward again.

However, I realize I had a different idea the other day so this one might also turn out to be a dead end too.

I have been trying (locally) to rewrite hash aggregate so it uses a future that was not created during the actual call to `poll_next`. However, I am currently stymied on the need to try and get `pin-ing` correct -- my branch refuses to compile. I'll keep posting updates....

Reply:
At this point I am ashamed of the immoral amount of time I spent on this 😭

I am getting convinced that I should have kept the other PR with the iterator. IMO that was easy to reason about (and there was already a 50% performance improvement xD

Reply:
It is fairly astonishing that Rust has made streaming so hard. 

However, I got it to compile and mostly work. Here is a commit (changes to this branch) with the approach that passes most of the the arrow and datafusion tests and also doesn't require waiting for all the input threads to complete:  https://github.com/apache/arrow/commit/28075970d8d59fa87359ad38470669485f029578

It still does not quite work in all cases (I need to port the normal aggregate over) but I'll finish that up.

I am curious if you have any thoughts about the approach. I'll try and write it up coherently tomorrow morning, now I need to go take care of some other things.

Reply:
you are in the right track IMO. Good thinking going for the Hash first.

Reply:
Here is the a commit that passes all the tests: https://github.com/alamb/arrow/commit/01b4d8758192b40494447401f8cb5bd970b38835 (again, based on this branch). 

What performance benchmarks were you running? I can see if this change makes any difference

Reply:
Cool! 

This is what I use:

```bash
git checkout master && \
    cargo bench --bench aggregate_query_sql && \
    git checkout merge && \
    cargo bench --bench aggregate_query_sql
```

Reply:
cool - thanks -- will give it a try tomorrow

Reply:
https://github.com/apache/arrow/pull/8553

Reply:
Super! I will close this one in favor of yours.